File: starter_kits/C#/Halite3/hlt/GameMap.cs
using System;
using System.Collections.Generic;

namespace Halite_Project.hlt {
    public class GameMap {
        public int width;
        public int height;
        public MapCell[][] cells;

        public GameMap(int width, int height) {
            this.width = width;
            this.height = height;

            cells = new MapCell[height][];
            for (int y = 0; y < height; ++y) {
                cells[y] = new MapCell[width];
            }
        }

        public MapCell At(Position position) {
            Position normalized = Normalize(position);
            return cells[normalized.y][normalized.x];
        }

        public MapCell At(Entity entity) {
            return At(entity.position);
        }

        public int CalculateDistance(Position source, Position target) {
            Position normalizedSource = Normalize(source);
            Position normalizedTarget = Normalize(target);

            int dx = Math.Abs(normalizedSource.x - normalizedTarget.x);
            int dy = Math.Abs(normalizedSource.y - normalizedTarget.y);

            int toroidal_dx = Math.Min(dx, width - dx);
            int toroidal_dy = Math.Min(dy, height - dy);

            return toroidal_dx + toroidal_dy;
        }

        public Position Normalize(Position position) {
            int x = ((position.x % width) + width) % width;
            int y = ((position.y % height) + height) % height;
            return new Position(x, y);
        }

        public List<EDirection> GetUnsafeMoves(Position source, Position destination) {
            List<EDirection> possibleMoves = new List<EDirection>();

            Position normalizedSource = Normalize(source);
            Position normalizedDestination = Normalize(destination);

            int dx = Math.Abs(normalizedSource.x - normalizedDestination.x);
            int dy = Math.Abs(normalizedSource.y - normalizedDestination.y);
            int wrapped_dx = width - dx;
            int wrapped_dy = height - dy;

            if (normalizedSource.x < normalizedDestination.x) {
                possibleMoves.Add(dx > wrapped_dx ? EDirection.WEST : EDirection.EAST);
            } else if (normalizedSource.x > normalizedDestination.x) {
                possibleMoves.Add(dx < wrapped_dx ? EDirection.WEST : EDirection.EAST);
            }

            if (normalizedSource.y < normalizedDestination.y) {
                possibleMoves.Add(dy > wrapped_dy ? EDirection.NORTH : EDirection.SOUTH);
            } else if (normalizedSource.y > normalizedDestination.y) {
                possibleMoves.Add(dy < wrapped_dy ? EDirection.NORTH : EDirection.SOUTH);
            }

            return possibleMoves;
        }

        public EDirection NaiveNavigate(Ship ship, Position destination) {
            // getUnsafeMoves normalizes for us
            foreach (EDirection direction in GetUnsafeMoves(ship.position, destination)) {
                Position targetPos = ship.position.DirectionalOffset(direction);
                if (!At(targetPos).IsOccupied()) {
                    At(targetPos).MarkUnsafe(ship);
                    return direction;
                }
            }

            return EDirection.STILL;
        }

        public void _update() {
            for (int y = 0; y < height; ++y) {
                for (int x = 0; x < width; ++x) {
                    cells[y][x].ship = null;
                }
            }

            int updateCount = Input.ReadInput().GetInt();

            for (int i = 0; i < updateCount; ++i) {
                Input input = Input.ReadInput();
                int x = input.GetInt();
                int y = input.GetInt();

                cells[y][x].halite = input.GetInt();
            }
        }

        public static GameMap _generate() {
            Input mapInput = Input.ReadInput();
            int width = mapInput.GetInt();
            int height = mapInput.GetInt();

            GameMap map = new GameMap(width, height);

            for (int y = 0; y < height; ++y) {
                Input rowInput = Input.ReadInput();

                for (int x = 0; x < width; ++x) {
                    int halite = rowInput.GetInt();
                    map.cells[y][x] = new MapCell(new Position(x, y), halite);
                }
            }

            return map;
        }
    }
}

File: starter_kits/Clojure/hlt/lib/hlt/game_map.clj
(ns hlt.game-map
  (:require [hlt.state :refer [game]]
            [hlt.input :refer [read-line-ints]]
            [hlt.direction :refer [invert]]
            [hlt.position :refer [new-position abs-position sub-position
                                  get-position directional-offset]]
            [hlt.map-cell :as map-cell :refer [read-map-cells]]
            [hlt.direction :as direction]))

(defn game-map
  "Return the game map"
  []
  (:game-map @game))

(defn at
  "Get cell at position"
  [position-or-entity]
  (get-in @game [:game-map :map-cells (get-position position-or-entity)]))

(defn is-occupied?
  "Is cell occupied?"
  [position-or-entity]
  (let [cell (at position-or-entity)]
    (map-cell/is-occupied? cell)))

(defn is-empty?
  "Is cell empty?"
  [position-or-entity]
  (let [cell (at position-or-entity)]
    (map-cell/is-empty? cell)))

(defn normalize
  "A method that normalizes a position within the bounds of the
  toroidal map. Useful for handling the wraparound modulus arithmetic
  on x and y. For example, if a ship at (x = 31, y = 4) moves to the
  east on a 32x32 map, the normalized position would be (x = 0, y =
  4), rather than the off-the-map position of (x = 32, y = 4)."
  [position]
  (let [[x y] position
        {:keys [width height]} (game-map)]
    [(mod x width) (mod y height)]))

(defn calculate-distance
  "A method that computes the Manhattan distance between two locations,
  and accounts for the toroidal wraparound."
  [source target]
  (let [source (-> source get-position normalize)
        target (-> target get-position normalize)
        [x y] (abs-position (sub-position source target))
        {:keys [width height]} (game-map)]
    (+ (min x (- width x)) (min y (- height y)))))

(defn get-target-direction
  "Returns where in the cardinality spectrum the target is from source.
  e.g.: North, East; South, West; etc."
  [source target]
  (let [[sx sy] (-> source get-position)
        [tx ty] (-> target get-position)
        cx (if (> tx sx)
             direction/east
             (if (< tx sx)
               direction/west))
        cy (if (> ty sy)
             direction/south
             (if (< ty sy)
               direction/north))]
    [cx cy]))

(defn get-unsafe-moves
  "A method that returns a list of direction(s) to move closer to a
  target disregarding collision possibilities. Returns an empty list
  if the source and destination are the same."
  [source destination]
  (let [{:keys [width height]} (game-map)
        source (-> source get-position normalize)
        destination (-> destination get-position normalize)
        [dx dy] (abs-position (sub-position destination source))
        [cx cy] (get-target-direction source destination)
        m0 (if-not (zero? dx)
             (if (< dx (/ width 2))
               cx
               (if cx (invert cx))))
        m1 (if-not (zero? dy)
             (if (< dy (/ height 2))
               cy
               (if cy (invert cy))))]
    (if m0
      (if m1
        [m0 m1]
        [m0])
      (if m1
        [m1]
        []))))

(defn naive-navigate
  "A method that returns a direction to move closer to a target
  without colliding with other entities. Returns a direction of
  still if no such move exists."
  [ship destination]
  (let [ship-position (get-position ship)
        destination-position (get-position destination)
        unsafe-moves (get-unsafe-moves ship-position destination-position)
        dir (reduce
              (fn [_ direction]
                (if (is-empty?
                      (directional-offset ship-position direction))
                  (reduced direction)))
              unsafe-moves)]
    (or dir direction/still)))

(defn read-game-map
  "Read the game map"
  []
  (let [[width height] (read-line-ints)
        map-cells (read-map-cells width height)]
    {:width width
     :height height
     :map-cells map-cells}))

(defn update-game-map
  [game-map]
  (let [{:keys [map-cells]} game-map
        remove-ship (fn [cells [position cell]]
                      (assoc cells
                        position (assoc cell
                                   :ship nil)))
        map-cells (reduce remove-ship
                    map-cells
                    (seq map-cells))
        [update-count] (read-line-ints)
        update-cell (fn [cells [x y halite]]
                      (let [position (new-position x y)
                            cell (get cells position)]
                        (assoc cells
                          position (assoc cell
                                     :halite halite))))
        map-cells (reduce update-cell
                    map-cells
                    (repeatedly update-count read-line-ints))]
    (assoc game-map
      :map-cells map-cells)))
